munit/shared/src/main/scala/munit/MUnitRunner.scala

```diff
@@ -144,80 +144,215 @@ class MUnitRunner(val cls: Class[_ <: Suite], newInstance: () => Suite)
     if (PlatformCompat.isIgnoreSuite(cls)) {
       val description = getDescription()
       notifier.fireTestIgnored(description)
-      return Future.successful(())
-    }
-    var isBeforeAllRun = false
-    val result = {
-      val isContinue = runBeforeAll(notifier)
-      isBeforeAllRun = isContinue
-      if (isContinue) {
-        runAsyncTestsSynchronously(notifier)
-      } else {
-        Future.successful(())
+      Future.successful(())
+    } else {
+      val result =
+        for {
+          isBeforeAllRun <- runBeforeAll(notifier)
+          if isBeforeAllRun
+          _ <- runAsyncTestsSynchronously(notifier)
+        } yield isBeforeAllRun
+
+      result.transformWithCompat {
+        case util.Success(true) =>
+          runAfterAll(notifier)
+
+        case util.Success(false) =>
+          Future.successful(())
+
+        case util.Failure(ex) =>
+          Future.failed(ex)
       }
     }
-    result.transformCompat { s =>
-      if (isBeforeAllRun) {
-        runAfterAll(notifier)
-      }
-      s
-    }
   }
 
-  private def runBeforeAll(notifier: RunNotifier): Boolean = {
-    var isContinue = runHiddenTest(notifier, "beforeAll", suite.beforeAll())
-    suite.munitFixtures.foreach { fixture =>
-      isContinue &= runHiddenTest(
-        notifier,
-        s"beforeAllFixture(${fixture.fixtureName})",
-        fixture.beforeAll()
+  private def runBeforeAll(notifier: RunNotifier): Future[Boolean] = {
+    val suiteBeforeAllF = Future(
+      runHiddenTest(notifier, "beforeAll", suite.beforeAll())
+    )
+
+    def beforeAllFixtureF =
+      Future.sequence(
```

Comment:
We should use a tail-recursive method to evaluate non-async futures in sequence so that stack traces doesn't get huge. If you're using a blocking `ExecutionContext` the stack trace will grow linearly with the number of test cases in the test suite. It should be possible to write test suites with ~100k test cases.

See `runAsyncTestsSynchronously` for how to do this. The trick is to have special handling when `future.value` is non-empty. You can't use the `@tailrec` annotation because the async path cannot be optimized for tail recursion, but the compiler still optimizes the sync path even if the annotation is not present.

https://github.com/scalameta/munit/blob/6484bf798c5fd3851e9b919964ab4d0837c57e94/munit/shared/src/main/scala/munit/MUnitRunner.scala#L127-L140

Reply:
Seems reasonable to me. I've added a similar to `runAsyncTestsSynchronously` recursion method for execution of a bunch of Futures.